Refine how resource classes with same components are merged

Postremus opened the pull request:
Fixes #25462 

Resources classes where merged by using all components as map key. This leads to situations where the only differing property of a component may be its name. The name is for a DEFAULT_REGEX however not enough to be considered a different path itself.

The merge is now based on a custom (String) key, build on properties which should enforce uniqueness when needed: type, literal, and pattern.

geoand commented:
Thanks for looking into this @Postremus! 

While working on this, please make sure that the JAX-RS TCK is passing with this change

Postremus commented:
@geoand
Where are the tcks located for resteasy-reactive? 
Is it this module? https://github.com/quarkusio/quarkus/tree/main/tcks/resteasy-reactive

geoand commented:
Yes exactly, Just run `mvn install` on that module once you've built your change in Quarkus.

Postremus committed:
Refine how resource classes with same components are merged
Resources classes where merged by using all components as map key. This leads to situations where the only differing property of a component may be its name. The name is for a DEFAULT_REGEX however not enough to be considered a different path itself.

The merge is now based on a custom (String) key, build on properties which should enforce uniqueness when needed: type, literal, and pattern.

Postremus commented:
@geoand 
TCKs are passing.

Postremus commented:
Test failures do not look related. The kafka dev service container could not be started.
```
java.lang.RuntimeException: 
java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.kafka.client.deployment.DevServicesKafkaProcessor#startKafkaDevService threw an exception: java.lang.RuntimeException: org.testcontainers.containers.ContainerLaunchException: Container startup failed
```

geoand reviewed (approved):
LGTM, thanks!

@stuartwdouglas any objections?

geoand reviewed (changes requested):
Actually, looking at https://github.com/quarkusio/quarkus/issues/25462#issuecomment-1126738354, it seems like this was an opt-in feature of RESTEasy Classic.

So we should probably do the same here.

Postremus commented:
I don't see how `resteasy.wider.request.matching` is related to this issue.

This flag was first mentioned, as far as I found, in the migration guide for 3.0-beta-5 to 3.0-beta-6:
https://docs.jboss.org/resteasy/docs/3.0.6.Final/userguide/html/Migration_from_older_versions.html
```
[..] For example, resource classes are scanned for a best match, other are ignored in the match. [..] There is one config switch I added so that Resteasy will ignore the Spec defined class expression filtering step and instead match base on the full expressions of each JAX-RS method. resteasy.wider.request.matching. [..]
```

I found these integration tests in resteasy:
* https://github.com/resteasy/resteasy/blob/main/testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/basic/WiderMappingTest.java
* https://github.com/resteasy/resteasy/blob/main/testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/basic/WiderMappingNegativeTest.java

The test consisist of:
* one resource class with path "/hello"
* one resource class with path "{x:.*}"

In WiderMappingTest (`resteasy.wider.request.matching=true`), {x:.*} must match, while "/hello" must match in the negative test.

Also, as far as I can tell, the handling of the flag `resteasy.wider.request.matching` is done in ResourceMethodRegistry. In each method of this class, `widerMatching` is checked. if true, do operations on `rootNode` (all endpoints), otherwise operate on `root` (only resource classes).

This flag only changes the order of how resources with different paths are merged, but has no influence in how resources with the same paths are matched.

I therefore believe that this fix is still correct, and no config switch is needed. We merge resources with the same path (components) on class level. No change is done to the specific resource matching.

Postremus commented:
I am btw. happy to introduce a config switch for this @geoand , I just don't want to do anything based on false assumptions on how the old stack worked.

stuartwdouglas reviewed (approved):
I think this looks ok.

geoand commented:
@Postremus thanks for this.

Any chance you can do a small benchmark to see if this introduces any noticeable overhead?

Postremus commented:
yep, will prepare a test project and do a simple benchmark (e.g. look at either profiler data or check how long dev mode restarts take comperativly)

geoand commented:
Thanks!

We also want to make sure throughput for simple hello-world style rest calls is not affected

gsmet commented on independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/startup/RuntimeDeploymentManager.java:
```diff
@@ -255,4 +260,54 @@ private String sanitizePathPrefix(String prefix) {
+                keyBuilder.append("|");
+            }
+
+            this.key = keyBuilder.toString();
```
My guess is that it would be better if you could compute the hashCode here and cache it in a final field.

gsmet reviewed (commented):
@Postremus could you have another look at this one so that we finalize it?

I added a small suggestion.

gsmet commented:
@Postremus friendly ping.

geoand commented:
@Postremus are you still interested in driving this home?

gsmet committed:
Apply some minor performance tweaks to MappersKey

This all comes from the observations I did in HV:

- Objects.hash() can be costly as it creates an array
- no real need to compare the classes given we know this object will
  only be used in a map that is consistent

Also:
- we don't need to cache the hashCode as String.hashCode() has a cache
- better use equals() than compareTo()

gsmet commented:
@geoand I rebased and applied some minor tweaks to reduce the cost of this new layer. I think you're more equipped than me to see if it introduces some significant overhead?

geoand commented:
I'll have a look

geoand committed:
Reduce allocations on RuntimeDeploymentManager.MappersKey

geoand commented:
Fixed the styling issue and applied one more small enhancement
